File: java/src/test/java/com/callisto/sdk/SmsTest.java
package com.callisto.sdk;

import com.callisto.sdk.models.Paginated;
import com.callisto.sdk.models.SendSmsResult;
import com.callisto.sdk.models.SmsMessage;
import com.callisto.sdk.resources.SmsSendRequest;
import com.fasterxml.jackson.databind.JsonNode;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class SmsTest {

    @Test
    void sendSingleRecipient() {
        StubHttpClient stub = new StubHttpClient(200,
                "{\"total_amount\": 5.0, \"available_credit\": 95.0, \"status\": \"queued\", "
                        + "\"recipient_count\": 1, \"scheduled\": false, \"messages\": []}");
        try (CallistoClient client = TestSupport.client(stub)) {
            SendSmsResult result = client.sms().send(SmsSendRequest.builder()
                    .sender("Acme").to("+225070").message("Hi").build());

            assertEquals("POST", stub.method());
            assertEquals("https://api.example.com/v1/sms/send", stub.uri().toString());
            JsonNode body = TestSupport.json(stub.lastBody);
            assertEquals("Acme", body.get("sender").asText());
            assertEquals("+225070", body.get("to").asText());
            assertEquals("Hi", body.get("message").asText());
            assertEquals("queued", result.getStatus());
            assertEquals(1, result.getRecipientCount());
        }
    }

    @Test
    void sendListWithOptionalFields() {
        StubHttpClient stub = new StubHttpClient(200,
                "{\"total_amount\": 0, \"available_credit\": 0, \"status\": \"queued\", "
                        + "\"recipient_count\": 2, \"scheduled\": true, \"messages\": []}");
        try (CallistoClient client = TestSupport.client(stub)) {
            client.sms().send(SmsSendRequest.builder()
                    .sender("Acme")
                    .to(List.of("+1", "+2"))
                    .message("Sale")
                    .notifyUrl("https://hook")
                    .scheduledAt("2026-06-02 10:00:00")
                    .build());

            JsonNode body = TestSupport.json(stub.lastBody);
            assertTrue(body.get("to").isArray());
            assertEquals(2, body.get("to").size());
            assertEquals("https://hook", body.get("notify_url").asText());
            assertEquals("2026-06-02 10:00:00", body.get("scheduled_at").asText());
        }
    }

    @Test
    void listBuildsQueryAndDecodesPagination() {
        StubHttpClient stub = new StubHttpClient(200,
                "{\"items\": [{\"id\": \"m1\", \"status\": \"sent\"}], \"total\": 1, "
                        + "\"per_page\": 50, \"current_page\": 1, \"next\": null, "
                        + "\"previous\": null, \"total_pages\": 1}");
        try (CallistoClient client = TestSupport.client(stub)) {
            Paginated<SmsMessage> page = client.sms().list(null, null, 1, 50);

            String uri = stub.uri().toString();
            assertTrue(uri.contains("page=1"), uri);
            assertTrue(uri.contains("per_page=50"), uri);
            assertEquals(1, page.getItems().size());
            assertEquals("m1", page.getItems().get(0).getId());
            assertEquals(1, page.getTotal());
            assertEquals(50, page.getPerPage());
            assertEquals(1, page.getCurrentPage());
            assertEquals(1, page.getTotalPages());
        }
    }

    @Test
    void getStatusUsesPathId() {
        StubHttpClient stub = new StubHttpClient(200, "{\"id\": \"abc\", \"status\": \"delivered\"}");
        try (CallistoClient client = TestSupport.client(stub)) {
            SmsMessage msg = client.sms().getStatus("abc");
            assertEquals("https://api.example.com/v1/sms/abc", stub.uri().toString());
            assertEquals("delivered", msg.getStatus());
        }
    }
}

File: js/test/sms.test.ts
import { describe, it, expect } from "vitest";
import { CallistoClient } from "../src/client.js";
import { mockFetch } from "./helpers.js";

const opts = { clientId: "cid", apiKey: "k", baseUrl: "https://api.test/v1" };

describe("sms", () => {
  it("send POSTs /sms/send with the body", async () => {
    const { fetch, captured } = mockFetch(200, {
      total_amount: 5, available_credit: 95, status: "sent",
      recipient_count: 1, scheduled: false, messages: [],
    });
    const client = new CallistoClient(opts, fetch);
    const res = await client.sms.send({ sender: "Acme", to: "+2250700000000", message: "Hi" });
    expect(captured[0].method).toBe("POST");
    expect(captured[0].url).toBe("https://api.test/v1/sms/send");
    expect(captured[0].body).toEqual({ sender: "Acme", to: "+2250700000000", message: "Hi" });
    expect(res.status).toBe("sent");
  });

  it("list GETs /sms/messages with pagination query", async () => {
    const message = {
      id: "abc", sender_name: "Acme", recipient: "+2250700000000",
      content: "Hi", status: "delivered",
      created_at: "2026-06-01 10:00:00", updated_at: "2026-06-01 10:01:00",
    };
    const { fetch, captured } = mockFetch(200, {
      items: [message], total: 1, per_page: 15, current_page: 1,
      next: 1, previous: 1, total_pages: 1,
    });
    const client = new CallistoClient(opts, fetch);
    const res = await client.sms.list({ page: 2, per_page: 50 });
    expect(captured[0].url).toBe("https://api.test/v1/sms/messages?page=2&per_page=50");
    expect(res.total).toBe(1);
    expect(res.items.length).toBe(1);
    expect(res.items[0].status).toBe("delivered");
    expect(res.items[0].sender_name).toBe("Acme");
  });

  it("getStatus GETs /sms/:id", async () => {
    const { fetch, captured } = mockFetch(200, {
      id: "abc", sender_name: "Acme", recipient: "+2250700000000",
      content: "Hi", status: "delivered",
      created_at: "2026-06-01 10:00:00", updated_at: "2026-06-01 10:01:00",
    });
    const client = new CallistoClient(opts, fetch);
    const res = await client.sms.getStatus("abc");
    expect(captured[0].url).toBe("https://api.test/v1/sms/abc");
    expect(res.status).toBe("delivered");
    expect(res.sender_name).toBe("Acme");
  });
});
